### PR TITLE
Introduced util for checking whether the editor wasn't initialized on the same element more than once

### PR DESCRIPTION
**Pull request by pomek**
### Suggested merge commit message ([convention](https://github.com/ckeditor/ckeditor5-design/wiki/Git-commit-message-convention))

Feature: Introduced a `secureSourceElement()` utility that makes sure source DOM elements are not shared between editor instances. See ckeditor/ckeditor5#746.

### CODE REVIEW

File: src/editor/utils/securesourceelement.js [hunk 1]
```diff
@@ -0,0 +1,47 @@
...
+import CKEditorError from '@ckeditor/ckeditor5-utils/src/ckeditorerror';
+
+/**
+ * @module core/editor/utils/securesourceelement
+ */
+
+const SECURE_ATTRIBUTE = 'data-ckeditor5';
+
+/**
+ * Checks if the editor was initialized using a source element. If yes, it prevents to creating another editor
```
Comment (jodator) [inline src/editor/utils/securesourceelement.js]: Could you rewrite this doc? The focus should be on "securing the source element" not on "checks if...".

File: src/editor/utils/securesourceelement.js [hunk 2]
```diff
@@ -0,0 +1,47 @@
...
+	const sourceElement = editor.sourceElement;
+
+	// If the editor was initialized without specifying an element, we don't need to secure anything.
+	if ( !sourceElement ) {
+		return;
+	}
+
+	if ( sourceElement.hasAttribute( SECURE_ATTRIBUTE ) ) {
+		/**
+		 * An element passed to the editor creator has been passed more than once. The element can be used only once.
```
Comment (jodator) [inline src/editor/utils/securesourceelement.js]:
```suggestion
		 * Another editor instance was already initialized on this element. You cannot create more then one editor on the same element.
```
Or something similar - I get the feeling that the error description should reflect how someone could use our API. So it is rather creating two or more instances of the ediotr on some element then passing something.

File: src/editor/utils/securesourceelement.js [hunk 3]
```diff
@@ -0,0 +1,47 @@
...
+		throw new CKEditorError(
+			'securesourceelement-source-element-used-more-than-once: ' +
+			'The editor cannot be initialized using the same source element more than once.'
+		);
+	}
+
+	// Mark the source element.
+	sourceElement.setAttribute( SECURE_ATTRIBUTE, 'true' );
+
+	// Remove the attribute when the editor is being destroyed.
```
Comment (jodator) [inline src/editor/utils/securesourceelement.js]: To explicit? You might omit this comment (and above also) - it is clear here that we remove this attribute on a "destroy" event. To go with pair with a comment above - it's about releasing the element so the source element can be used again. But I think that's clear here :)

File: tests/editor/utils/securesourceelement.js [hunk 4]
```diff
@@ -0,0 +1,69 @@
...
+		return editor.destroy()
+			.then( () => {
+				editor = null;
+
+				expect( sourceElement.hasAttribute( 'data-ckeditor5' ) ).to.equal( false );
+			} );
+	} );
+
+	it( 'throws an error if the same element was used twice', () => {
+		sourceElement.setAttribute( 'data-ckeditor5', 'true' );
```
Comment (jodator) [inline tests/editor/utils/securesourceelement.js]:
```suggestion
		secureSourceElement( editor );
```
I'd go with explicitly calling the method twice here.

File: src/editor/utils/securesourceelement.js [hunk 5]
```diff
@@ -0,0 +1,47 @@
...
+	}
+
+	if ( sourceElement.hasAttribute( SECURE_ATTRIBUTE ) ) {
+		/**
+		 * An element passed to the editor creator has been passed more than once. The element can be used only once.
+		 *
+		 * @error securesourceelement-source-element-used-more-than-once
+		 */
+		throw new CKEditorError(
+			'securesourceelement-source-element-used-more-than-once: ' +
```
Comment (jodator) [inline src/editor/utils/securesourceelement.js]: I'm not sure about this name. @Reinmar - wouldn't be nice something different?

1. The prefix doesn't tell much. I'd go with `editor-` as other similar errors like [editor-missing-sourceelement](https://ckeditor.com/docs/ckeditor5/latest/framework/guides/support/error-codes.html#error-editor-missing-sourceelement).
2. The error name itself is long. `editor-sourceelemnt-already-used` or `editor-sourcelement-in-use`. But maybe `editor-sourceelement-used-more-than-once` is OK :thinking:.

**Review by jodator — CHANGES_REQUESTED** [review]
Some updates to docs are needed - the PR is OK :) I have only one question to @Reinmar about naming the error.

File: src/editor/utils/securesourceelement.js (hunk 3 above)
Comment (Reinmar) [reply src/editor/utils/securesourceelement.js]: The same for the previous line, in fact.

File: src/editor/utils/securesourceelement.js (hunk 5 above)
Comment (Reinmar) [reply src/editor/utils/securesourceelement.js]: `editor-source-element-already-used` sounds fine to me.

**Comment by oleq** [comment]
Why not a part of [`ElementApiMixin`](https://ckeditor.com/docs/ckeditor5/latest/api/module_core_editor_utils_elementapimixin-ElementApiMixin.html#member-sourceElement)?

**Comment by jodator** [comment]
> Why not a part of [`ElementApiMixin`](https://ckeditor.com/docs/ckeditor5/latest/api/module_core_editor_utils_elementapimixin-ElementApiMixin.html#member-sourceElement)?

~I think that probably for the same reason as https://github.com/ckeditor/ckeditor5-core/pull/173#discussion_r286366282~

~I think that not every editor build is initialized over an element.~

Edit: I was wrong - @oleq this looks like a place for this util.

**Comment by oleq** [comment]
I moved the method and updated other PRs. 

**But don't merge it please**, because I figured we can go further and have `ElementDataMixin#setSourceElement()`, that will handle these

```js
if ( isElement( sourceElementOrData ) ) {
	this.sourceElement = sourceElementOrData;
	this.secureSourceElement();
}
```

in editor constructors as 

```js
if ( isElement( sourceElementOrData ) ) {
	this.setSourceElement( sourceElementOrData );
}
```

with backwards compatibility (setting `this.sourceElement` directly will warn about this way being obsolete like we did with [`EditorUI.editableElements`](https://github.com/ckeditor/ckeditor5-core/blob/t/ckeditor5/746/src/editor/editorui.js#L156-L171)).

Gonna create it ASAP.

**Comment by oleq** [comment]
I reverted some of the changes I made because it turned out the `ElementApi` interface [does not make sense the way we use it](https://github.com/ckeditor/ckeditor5-core/issues/189) (RFC). Securing the editor element should remain a helper. The code is ready for review.

**Comment by oleq** [comment]
Self–review: maybe instead of using `data-` attributes we should simply use DOM element properties? Less clutter in the view/DOM.

**Comment by jodator** [comment]
> Self–review: maybe instead of using `data-` attributes we should simply use DOM element properties? Less clutter in the view/DOM.

Sounds cool :)

**Comment by Reinmar** [comment]
> Self–review: maybe instead of using `data-` attributes we should simply use DOM element properties? Less clutter in the view/DOM.

Just like `editableElement.ckeditorInstance`? In fact, we could assign `ckeditorInstance` there for the same reasons why we marked editable elements with it :)

**Comment by oleq** [comment]
In fact... we could use [`#ckeditorInstance`](https://github.com/ckeditor/ckeditor5-core/blob/810e4a4a0c1b5242f694ba6a7b20d6dd1471473b/src/editor/editorui.js#L127) that we recently implemented.

**Comment by oleq** [comment]
Perfect timing.

File: src/editor/utils/securesourceelement.js [hunk 6]
```diff
@@ -0,0 +1,47 @@
...
+			'The editor cannot be initialized using the same source element more than once.'
+		);
```
Comment (ma2ciek) [inline src/editor/utils/securesourceelement.js:42]: The error context is missing (`editor`).

File: tests/editor/utils/securesourceelement.js [hunk 7]
```diff
@@ -0,0 +1,69 @@
...
+
+		expect( () => {
```
Comment (ma2ciek) [inline tests/editor/utils/securesourceelement.js]: You should now use the `expectToThrowCKEditorError` helper. It also allows checking if the error context is reachable from the editor.

File: src/editor/utils/securesourceelement.js (hunk 6 above)
Comment (oleq) [reply src/editor/utils/securesourceelement.js:42]: You're reviewing an old code :)

File: tests/editor/utils/securesourceelement.js (hunk 7 above)
Comment (oleq) [reply tests/editor/utils/securesourceelement.js]: Again, this is the old source.

**Comment by oleq** [comment]
Ready for another round of review.

**Review by jodator — APPROVED** [review]
:+1: I've fixed only one docs as I couldn't parse it :)

**Pull request merged**